### PR TITLE
Remote Index Handles 

### PR DESCRIPTION
**Pull request by hntd187**
The idea here is that when toshi spins up, it determines if it's the master or not. If it's the master it creates node clients. If it's not the master they spin up grpc services and not http services. The catalog is then built on a per-node basis, data nodes _should_ only have `LocalIndexHandles` where as the master should/could have a mix of `RemoteIndexHandle` and `LocalIndexHandle` depending on where the pieces of the index actually lay.

### CODE REVIEW

File: src/handle.rs [hunk 1]
```diff
@@ -1,19 +1,151 @@
...
+use crate::query::{CreateQuery, Query, Request};
+use crate::results::{ScoredDoc, SearchResults};
 use crate::settings::Settings;
 use crate::Result;
 
-use std::sync::atomic::{AtomicUsize, Ordering};
-use std::sync::{Arc, Mutex};
-use tantivy::{Index, IndexWriter};
+pub enum IndexLocation {
+    LOCAL,
```
Comment (LucioFranco) [inline src/handle.rs]: I believe that generally enums are Pascal case and constants are full caps.

File: src/handle.rs [hunk 2]
```diff
@@ -1,19 +1,151 @@
...
 }
 
-impl IndexHandle {
+impl IndexHandle for LocalIndexHandle {
+    fn get_name(&self) -> String {
+        self.name.clone()
+    }
+
+    fn index_location(&self) -> IndexLocation {
+        IndexLocation::LOCAL
```
Comment (LucioFranco) [inline src/handle.rs]: Might be interesting to use traits to enforce that we only use local or remote ones at compile time.

I think we could do this via marker traits that don't have any impl details.

File: src/cluster/remote_handle.rs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+use crate::handle::{IndexHandle, IndexLocation};
+use crate::handlers::index::{AddDocument, DeleteDoc, DocsAffected};
+use crate::query::Request;
+use crate::results::SearchResults;
+use crate::settings::Settings;
+use crate::Result;
+
+/// A reference to an index stored somewhere else on the cluster, this operates via calling
+/// the remote host and full filling the request via rpc
+pub struct RemoteIndexHandle<C> {
```
Comment (LucioFranco) [inline src/cluster/remote_handle.rs]: I think this name is getting a bit too long I think we should think of refactoring that.

I actually think we should just drop the `XXXIndexHandle` pattern and leave it as `RemoteIndex` and `LocalIndex`.

File: src/handle.rs (hunk 1 above)
Comment (hntd187) [reply src/handle.rs]: Just another reason you can tell I come from the JVM world https://docs.oracle.com/javase/tutorial/java/javaOO/enum.html

File: src/cluster/remote_handle.rs [hunk 4]
```diff
@@ -1,22 +1,42 @@
 #![allow(dead_code, unused_variables)]
 
+use std::io::Error as IOError;
```
Comment (LucioFranco) [inline src/cluster/remote_handle.rs]: very small nit but I prefer to use `io::Error` over `IOError`

File: src/cluster/remote_handle.rs [hunk 5]
```diff
@@ -57,7 +62,13 @@ impl IndexHandle for RemoteIndex {
             .and_then(move |mut client| {
                 let bytes = serde_json::to_vec(&search).unwrap();
                 let req = TowerRequest::new(SearchRequest { index: name, query: bytes });
-                client.search_index(req).map(|res| res.into_inner()).map_err(|_| Error::Inner(()))
+                client
+                    .search_index(req)
+                    .map(|res| {
+                        println!("{:#?}", res);
```
Comment (LucioFranco) [inline src/cluster/remote_handle.rs]: probably want to drop this println or change it for a `debug!`

**Pull request merged**